sae_lens/sae.py

```diff
@@ -397,109 +399,23 @@ def forward(
-                hidden_pre = sae_in_cent @ self.W_enc + self.b_enc
-                feature_acts = self.activation_fn(hidden_pre)
-                x_reconstruct_clean = self.reshape_fn_out(
-                    self.apply_finetuning_scaling_factor(feature_acts) @ self.W_dec
-                    + self.b_dec,
-                    d_head=self.d_head,
-                )
-
-                sae_out = self.run_time_activation_norm_fn_out(sae_out)
-                sae_error = self.hook_sae_error(x - x_reconstruct_clean)
-            return self.hook_sae_output(sae_out + sae_error)
-
-        # TODO: Add tests
-        elif self.use_error_term and self.cfg.architecture == "gated":
-            with torch.no_grad():
-                x = x.to(self.dtype)
-                sae_in = self.reshape_fn_in(x)  # type: ignore
-
-                # handle run time activation normalization if needed
-                sae_in = self.run_time_activation_norm_fn_in(sae_in)
-
-                # apply b_dec_to_input if using that method.
-                sae_in = sae_in - (self.b_dec * self.cfg.apply_b_dec_to_input)
-
-                gating_pre_activation = sae_in @ self.W_enc + self.b_gate
-                active_features = (gating_pre_activation > 0).float()
-
-                # Magnitude path with weight sharing
-                magnitude_pre_activation = self.hook_sae_acts_pre(
-                    sae_in @ (self.W_enc * self.r_mag.exp()) + self.b_mag
-                )
-                feature_magnitudes = self.activation_fn(magnitude_pre_activation)
-                feature_acts_clean = self.hook_sae_acts_post(
-                    active_features * feature_magnitudes
-                )
-                x_reconstruct_clean = self.reshape_fn_out(
-                    self.apply_finetuning_scaling_factor(feature_acts_clean)
-                    @ self.W_dec
-                    + self.b_dec,
-                    d_head=self.d_head,
-                )
-
+                with _disable_hooks(self):
+                    feature_acts_clean = self.encode(x)
+                    x_reconstruct_clean = self.decode(feature_acts_clean)
                 sae_error = self.hook_sae_error(x - x_reconstruct_clean)
-            return self.hook_sae_output(sae_out + sae_error)
-
-        # TODO: Add tests
-        elif self.use_error_term and self.cfg.architecture == "jumprelu":
-            with torch.no_grad():
-                x = x.to(self.dtype)
-                sae_in = self.reshape_fn_in(x)  # type: ignore
-
-                # handle run time activation normalization if needed
-                sae_in = self.run_time_activation_norm_fn_in(sae_in)
-
-                # apply b_dec_to_input if using that method.
-                sae_in = sae_in - (self.b_dec * self.cfg.apply_b_dec_to_input)
-
-                # "... d_in, d_in d_sae -> ... d_sae",
-                hidden_pre = sae_in @ self.W_enc + self.b_enc
-                feature_acts = self.hook_sae_acts_post(
```

Comment:
calling the hook here is a bug and is responsible for #326

Reply:
correct, thanks for catching this!